### ISSUE TITLE
Memory Issue when using Multidataset + AIStore + Horovod

### ISSUE THREAD

**Issue by xb478** [tmbdev/webdataset#22]
Problem description:
Possible memory leak / free up memory issue or multi processing issue when using Multidataset instead of the pytorch Dataloader.
Setup:
AIstore locally as the storage backend.
Webdataset:
``` 
url = "{}{}-{{000000..{:06d}}}.tar".format(
            ais_path, self._config['Datahandler']['name']['train'], num_shards)
url = f"pipe:curl -L -s {url} || true"
...
_ds = wds.Dataset(url, length = num_batches, shard_selection = worker_urls)\
                              .shuffle(self._config['Datahandler']['shuffle_buffer_size'])\
                              .decode('pil')\
                              .to_tuple("jpg;png;pyd;txt;ppm", "json;cls")\
                              .map_tuple(self._model_module.transform_train, self._model_module.transform_target)
...
dataloader = wds.MultiDataset(_ds, workers = num_workers, nominal=nominal).batched(batch_size)
```
Training backend: PyTorch Lightning with horovod as ddp strategy.
Num workers: 4/gpu
Num Gpus: 2-16 (configurable)

So what happens is that the training starts and the Multidatasets are setup in each of horovod/mpirun spawned processes according to the configuration. Then the memory starts filling up and I am able to identify that somehow there is a whole bunch of python processes (of the script that does the training and which is called via horovodrun -np <numgpus> python <path_to_entry_point.py>) that do not seem to finish but each occupy around 1-2 GB of memory.
They never seem to be joined/ finished and hence eventually the memory of the machine will be full after which swap will be filled and the whole training process obviously stalls.

I tried changing the order of calling shuffling as shown in the webdataset repository unittest of the Multidataset but it did not change the behavior. 

Is there something I am missing here? Btw. training with ResizedDataset and the PyTorch dataloader works fine even though the Dataloader throws a bunch of warnings because of inconsistent dataset lengths returned (but I think you mention this strange behavior in the Readme and thats why there is Multidataset).

**Comment by tmbdev** [comment]
I believe this was fixed here: https://github.com/tmbdev/webdataset/pull/18

Please note that `MultiDataset` is entirely optional; WebDataset works fine with the regular `DataLoader`.

Furthermore, if you want more scalability and easier testing/debugging, you might want to consider using github.com/nvlabs/tensorcom

**Comment by g1910** [comment]
I am facing a similar issue using WebDataset. It seems that over time the memory is completely filled up with loading the shards. What is the way to close/remove open shards from memory?

**Comment by nauyan** [comment]
@g1910 any fix for the above issue?

**Comment by tmbdev** [comment]
I haven't identified any storage leaks. If you can find a source with a memory profiler, I can try to fix it if it is an actual bug.

I've been working on a Ray Data port of webdataset, which gives you a more principled way of running distributed multicore preprocessing. There will likely also be a UCX-based loader.
